File: 2.link-list/5.get-node-value/get-node-value.go
// Write a function, getNodeValue, that takes in the head of a linked list and an index. The function should return the value of the linked list at the specified index.

// If there is no node at the given index, then return null.

package main

import "fmt"

type Node struct {
	value string
	next  *Node
}

func NewNode(value string) *Node {
	return &Node{value: value}
}

// iterative
// func getNodeValue(head *Node, index int) *string {
// 	for head != nil {
// 		if index == 0 {
// 			return &head.value
// 		}
// 		head = head.next
// 		index--
// 	}
// 	return nil
// }

// recursive
func getNodeValue(head *Node, index int) *string {
	if head == nil {
		return nil
	}
	if index == 0 {
		return &head.value
	}
	return getNodeValue(head.next, index-1)
}

func main() {
	a := NewNode("a")
	b := NewNode("b")
	c := NewNode("c")
	d := NewNode("d")

	a.next = b
	b.next = c
	c.next = d

	if result := getNodeValue(a, 2); result != nil {
		fmt.Println(*result)
	} else {
		fmt.Println(result)
	}
}

File: link-list/get-node-value/get-node-value.js
// Write a function, getNodeValue, that takes in the head of a linked list and an index. The function should return the value of the linked list at the specified index.

// If there is no node at the given index, then return null.

class Node {
    constructor(val) {
        this.val = val;
        this.next = null;
    }
}

const getNodeValue = (head, index) => {
    while (head !== null) {
        if (index === 0) {
            return head.val
        }
        head = head.next
        index--
    }
    return null
};

const a = new Node("a");
const b = new Node("b");
const c = new Node("c");
const d = new Node("d");

a.next = b;
b.next = c;
c.next = d;

// a -> b -> c -> d

console.log(getNodeValue(a, 2)); // 'c'